File: app (1).py
"""
app.py
------
FastAPI backend for the Urumuri Pharma Ltd analytics dashboard.

Reads from data/sales.db (SQLite) and exposes aggregated JSON
endpoints the frontend dashboard consumes. Most endpoints accept an
optional `?months=6` query param — pass it to filter the underlying
sales data to the most recent N months; omit it for all-time (12 months
in this dataset).

Run:
    pip install -r requirements.txt
    python3 data_generator.py        # generates data/sales.db (run once)
    uvicorn app:app --reload --port 8000

Then open frontend/index.html in a browser — it fetches from
http://localhost:8000/api/* and falls back to frontend/data.json if
the API isn't reachable.
"""

import sqlite3
from pathlib import Path
from typing import Optional

from fastapi import FastAPI, HTTPException, Query
from fastapi.middleware.cors import CORSMiddleware

from insights import compute_insights, get_cutoff_date

BASE_DIR = Path(__file__).resolve().parent.parent
DB_PATH = BASE_DIR / "data" / "sales.db"
MONTHLY_CEILING_RWF = 15_000_000

app = FastAPI(
    title="Urumuri Pharma Analytics API",
    description="Sales, inventory, and insights for a Rwandan pharmaceutical distribution company.",
    version="1.1.0",
)

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_methods=["*"],
    allow_headers=["*"],
)


def get_conn():
    if not DB_PATH.exists():
        raise HTTPException(
            status_code=500,
            detail="Database not found. Run 'python3 data_generator.py' first.",
        )
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    return conn


def cutoff_for(conn, months: Optional[int]):
    if not months:
        return None
    cur = conn.cursor()
    cur.execute("SELECT MAX(date) AS d FROM sales")
    max_date = cur.fetchone()["d"]
    return get_cutoff_date(max_date, months)


def sales_rows(conn, cutoff):
    cur = conn.cursor()
    if cutoff:
        cur.execute("SELECT * FROM sales WHERE date >= ?", (cutoff,))
    else:
        cur.execute("SELECT * FROM sales")
    return [dict(r) for r in cur.fetchall()]


@app.get("/")
def root():
    return {
        "service": "Urumuri Pharma Analytics API",
        "currency": "RWF",
        "endpoints": [
            "/api/kpis?months=6",
            "/api/monthly-revenue?months=6",
            "/api/category-breakdown?months=6",
            "/api/region-breakdown?months=6",
            "/api/client-type-breakdown?months=6",
            "/api/top-products?months=6&limit=8",
            "/api/new-orders?limit=12",
            "/api/inventory?months=6&status=all|reorder",
            "/api/insights?months=6",
        ],
    }


@app.get("/api/kpis")
def kpis(months: Optional[int] = Query(None)):
    conn = get_conn()
    cutoff = cutoff_for(conn, months)
    rows = sales_rows(conn, cutoff)

    total_orders = len(rows)
    total_revenue = sum(r["line_total_rwf"] for r in rows)

    monthly = {}
    for r in rows:
        monthly.setdefault(r["month"], 0)
        monthly[r["month"]] += r["line_total_rwf"]
    best_month = max(monthly, key=monthly.get) if monthly else None

    regions = {r["region"] for r in rows}
    products = {r["product"] for r in rows}

    inv = inventory_rows(conn, cutoff)
    reorder_count = sum(1 for r in inv if r["needs_reorder"])

    conn.close()
    return {
        "total_revenue_rwf": total_revenue,
        "total_orders": total_orders,
        "avg_order_value_rwf": round(total_revenue / total_orders) if total_orders else 0,
        "best_month": best_month,
        "best_month_revenue_rwf": monthly.get(best_month, 0),
        "active_regions": len(regions),
        "active_products": len(products),
        "reorder_count": reorder_count,
        "monthly_ceiling_rwf": MONTHLY_CEILING_RWF,
    }


@app.get("/api/monthly-revenue")
def monthly_revenue(months: Optional[int] = Query(None)):
    conn = get_conn()
    cutoff = cutoff_for(conn, months)
    rows = sales_rows(conn, cutoff)
    conn.close()
    monthly = {}
    for r in rows:
        monthly.setdefault(r["month"], 0)
        monthly[r["month"]] += r["line_total_rwf"]
    return [{"month": m, "revenue_rwf": monthly[m]} for m in sorted(monthly)]


@app.get("/api/category-breakdown")
def category_breakdown(months: Optional[int] = Query(None)):
    conn = get_conn()
    cutoff = cutoff_for(conn, months)
    rows = sales_rows(conn, cutoff)
    conn.close()
    totals = {}
    for r in rows:
        totals.setdefault(r["category"], 0)
        totals[r["category"]] += r["line_total_rwf"]
    return sorted(
        ({"category": k, "revenue_rwf": v} for k, v in totals.items()),
        key=lambda x: x["revenue_rwf"], reverse=True,
    )


@app.get("/api/region-breakdown")
def region_breakdown(months: Optional[int] = Query(None)):
    conn = get_conn()
    cutoff = cutoff_for(conn, months)
    rows = sales_rows(conn, cutoff)
    conn.close()
    totals = {}
    for r in rows:
        totals.setdefault(r["region"], 0)
        totals[r["region"]] += r["line_total_rwf"]
    return sorted(
        ({"region": k, "revenue_rwf": v} for k, v in totals.items()),
        key=lambda x: x["revenue_rwf"], reverse=True,
    )


@app.get("/api/client-type-breakdown")
def client_type_breakdown(months: Optional[int] = Query(None)):
    conn = get_conn()
    cutoff = cutoff_for(conn, months)
    rows = sales_rows(conn, cutoff)
    conn.close()
    totals = {}
    for r in rows:
        totals.setdefault(r["client_type"], 0)
        totals[r["client_type"]] += r["line_total_rwf"]
    return sorted(
        ({"client_type": k, "revenue_rwf": v} for k, v in totals.items()),
        key=lambda x: x["revenue_rwf"], reverse=True,
    )


@app.get("/api/top-products")
def top_products(months: Optional[int] = Query(None), limit: int = 8):
    conn = get_conn()
    cutoff = cutoff_for(conn, months)
    rows = sales_rows(conn, cutoff)
    conn.close()
    totals = {}
    for r in rows:
        totals.setdefault(r["product"], {"revenue": 0, "units": 0})
        totals[r["product"]]["revenue"] += r["line_total_rwf"]
        totals[r["product"]]["units"] += r["quantity"]
    ranked = sorted(
        ({"product": k, **v} for k, v in totals.items()),
        key=lambda x: x["revenue"], reverse=True,
    )
    return ranked[:limit]


@app.get("/api/new-orders")
def new_orders(limit: int = 12):
    """Most recent transactions, newest first — shown as the
    'New Orders' panel near the top of the dashboard."""
    conn = get_conn()
    cur = conn.cursor()
    cur.execute("""
        SELECT transaction_id, date, product, category, quantity,
               line_total_rwf, region, client_name, client_type
        FROM sales ORDER BY date DESC, transaction_id DESC LIMIT ?
    """, (limit,))
    rows = [dict(r) for r in cur.fetchall()]
    conn.close()
    return rows


def inventory_rows(conn, cutoff):
    """Inventory joined with sales velocity for the given window, sorted
    by urgency (needs_reorder first, then soonest to run out) — not
    alphabetical."""
    cur = conn.cursor()
    if cutoff:
        cur.execute("""
            SELECT product, SUM(quantity) AS units, MIN(date) AS d0, MAX(date) AS d1
            FROM sales WHERE date >= ? GROUP BY product
        """, (cutoff,))
    else:
        cur.execute("""
            SELECT product, SUM(quantity) AS units, MIN(date) AS d0, MAX(date) AS d1
            FROM sales GROUP BY product
        """)
    velocity = {}
    for row in cur.fetchall():
        from datetime import date as _date
        d0, d1 = _date.fromisoformat(row["d0"]), _date.fromisoformat(row["d1"])
        span = max(1, (d1 - d0).days + 1)
        velocity[row["product"]] = row["units"] / span

    cur.execute("SELECT * FROM inventory")
    items = []
    for row in cur.fetchall():
        d = dict(row)
        avg_daily = velocity.get(d["product"], 0)
        d["avg_daily_units_sold"] = round(avg_daily, 2)
        d["days_of_stock_remaining"] = round(d["stock_on_hand"] / avg_daily, 1) if avg_daily > 0 else None
        d["needs_reorder"] = d["stock_on_hand"] <= d["reorder_threshold"]
        items.append(d)

    items.sort(key=lambda r: (
        0 if r["needs_reorder"] else 1,
        r["days_of_stock_remaining"] if r["days_of_stock_remaining"] is not None else 9_999,
    ))
    return items


@app.get("/api/inventory")
def inventory(months: Optional[int] = Query(None), status: str = "all"):
    conn = get_conn()
    cutoff = cutoff_for(conn, months)
    items = inventory_rows(conn, cutoff)
    conn.close()
    if status == "reorder":
        items = [i for i in items if i["needs_reorder"]]
    return items


@app.get("/api/insights")
def insights(months: Optional[int] = Query(None)):
    conn = get_conn()
    cutoff = cutoff_for(conn, months)
    rows = sales_rows(conn, cutoff)
    inv = inventory_rows(conn, cutoff)
    conn.close()

    monthly, category_totals, region_totals = {}, {}, {}
    for r in rows:
        monthly.setdefault(r["month"], 0)
        monthly[r["month"]] += r["line_total_rwf"]
        category_totals.setdefault(r["category"], 0)
        category_totals[r["category"]] += r["line_total_rwf"]
        region_totals.setdefault(r["region"], 0)
        region_totals[r["region"]] += r["line_total_rwf"]

    total_revenue = sum(monthly.values())
    return compute_insights(monthly, category_totals, region_totals, total_revenue, len(rows), inv)

File: app (1).js
/* ===========================================================
   app.js — Urumuri Pharma Analytics Dashboard
   Tries the FastAPI backend first; if it isn't running, falls
   back to the static data.json snapshot so this still works
   when you just double-click index.html.
   =========================================================== */

const API_BASE = "http://localhost:8000/api";

const PALETTE = {
  gold: "#D9A130",
  goldDim: "#B4862C",
  sage: "#7FB69E",
  sageDim: "#5C8C77",
  clay: "#E2725B",
  teal600: "#246B61",
  teal700: "#18524B",
  cream: "#F2EFE6",
  creamDim: "#C9C7BC",
};

const CATEGORY_COLORS = [
  PALETTE.gold, PALETTE.sage, PALETTE.clay, PALETTE.teal600,
  PALETTE.goldDim, PALETTE.sageDim, PALETTE.teal700, PALETTE.creamDim,
  "#8C6BB1", "#4F8A8B",
];

function fmtRWF(n, opts = {}) {
  const rounded = Math.round(n || 0);
  const str = rounded.toLocaleString("en-US");
  return opts.short ? `RWF ${str}` : `RWF ${str}`;
}

function fmtShort(n) {
  if (n >= 1_000_000) return `${(n / 1_000_000).toFixed(1)}M`;
  if (n >= 1_000) return `${(n / 1_000).toFixed(0)}K`;
  return `${n}`;
}

function fmtDate(iso) {
  const d = new Date(iso + "T00:00:00");
  return d.toLocaleDateString("en-GB", { day: "2-digit", month: "short" });
}

async function fetchJSON(url) {
  const res = await fetch(url, { cache: "no-store" });
  if (!res.ok) throw new Error(`${url} -> ${res.status}`);
  return res.json();
}

async function loadData() {
  // Try the live API first.
  try {
    const [kpis, monthly, category, region, clientType, top, recent] = await Promise.all([
      fetchJSON(`${API_BASE}/kpis`),
      fetchJSON(`${API_BASE}/monthly-revenue`),
      fetchJSON(`${API_BASE}/category-breakdown`),
      fetchJSON(`${API_BASE}/region-breakdown`),
      fetchJSON(`${API_BASE}/client-type-breakdown`),
      fetchJSON(`${API_BASE}/top-products?limit=8`),
      fetchJSON(`${API_BASE}/recent-transactions?limit=12`),
    ]);
    setConnectionLabel(true);
    return {
      kpis,
      monthly_revenue: monthly,
      category_breakdown: category,
      region_breakdown: region,
      client_type_breakdown: clientType,
      top_products: top.map(r => ({ product: r.product, revenue: r.revenue, units: r.units })),
      recent_transactions: recent.map(r => ({
        date: r.date, client_name: r.client_name, product: r.product,
        region: r.region, line_total_rwf: r.line_total_rwf,
      })),
    };
  } catch (err) {
    console.warn("API not reachable, falling back to data.json snapshot.", err);
    setConnectionLabel(false);
    const snapshot = await fetchJSON("data.json");
    return snapshot;
  }
}

function setConnectionLabel(live) {
  const label = document.getElementById("connection-label");
  const dot = document.querySelector(".dot--live");
  if (live) {
    label.textContent = "Live API connected";
  } else {
    label.textContent = "Offline mode — showing saved snapshot";
    dot.style.background = PALETTE.creamDim;
    dot.style.animation = "none";
  }
}

function renderKPIs(d) {
  const ceiling = d.kpis.monthly_ceiling_rwf || 15_000_000;
  const pctOfCeiling = Math.round((d.kpis.best_month_revenue_rwf / ceiling) * 100);

  const cards = [
    {
      label: "Total revenue (YTD)",
      value: `RWF ${fmtShort(d.kpis.total_revenue_rwf)}`,
      sub: fmtRWF(d.kpis.total_revenue_rwf),
    },
    {
      label: "Total orders",
      value: d.kpis.total_orders.toLocaleString("en-US"),
      sub: `Avg order ${fmtRWF(d.kpis.avg_order_value_rwf)}`,
    },
    {
      label: "Best month",
      value: d.kpis.best_month,
      sub: `${fmtRWF(d.kpis.best_month_revenue_rwf)} · ${pctOfCeiling}% of ceiling`,
      warn: pctOfCeiling >= 95,
    },
    {
      label: "Coverage",
      value: `${d.kpis.active_regions} regions`,
      sub: `${d.kpis.active_products} active SKUs`,
    },
  ];

  const row = document.getElementById("kpi-row");
  row.innerHTML = cards.map(c => `
    <div class="kpi">
      <p class="kpi__label">${c.label}</p>
      <p class="kpi__value">${c.value}</p>
      <p class="kpi__sub ${c.warn ? "kpi__sub--warn" : ""}">${c.sub}</p>
    </div>
  `).join("");
}

function renderMonthlyChart(d) {
  const ctx = document.getElementById("chart-monthly");
  const labels = d.monthly_revenue.map(r => r.month);
  const values = d.monthly_revenue.map(r => r.revenue_rwf);
  const ceiling = d.kpis.monthly_ceiling_rwf || 15_000_000;

  new Chart(ctx, {
    type: "line",
    data: {
      labels,
      datasets: [
        {
          label: "Revenue",
          data: values,
          borderColor: PALETTE.gold,
          backgroundColor: "rgba(217,161,48,0.18)",
          pointBackgroundColor: PALETTE.gold,
          pointBorderColor: PALETTE.teal900 || "#0E3B36",
          fill: true,
          tension: 0.35,
          borderWidth: 2.5,
          pointRadius: 3,
        },
        {
          label: "Monthly ceiling (RWF 15,000,000)",
          data: labels.map(() => ceiling),
          borderColor: PALETTE.clay,
          borderDash: [6, 6],
          borderWidth: 1.5,
          pointRadius: 0,
          fill: false,
        },
      ],
    },
    options: {
      responsive: true,
      maintainAspectRatio: false,
      interaction: { mode: "index", intersect: false },
      plugins: {
        legend: { display: false },
        tooltip: {
          backgroundColor: "#0A2826",
          borderColor: "rgba(242,239,230,0.15)",
          borderWidth: 1,
          titleFont: { family: "IBM Plex Mono" },
          bodyFont: { family: "IBM Plex Mono" },
          callbacks: {
            label: (ctx) => `${ctx.dataset.label}: ${fmtRWF(ctx.parsed.y)}`,
          },
        },
      },
      scales: {
        x: {
          grid: { color: "rgba(242,239,230,0.06)" },
          ticks: { color: PALETTE.creamDim, font: { family: "IBM Plex Mono", size: 11 } },
        },
        y: {
          grid: { color: "rgba(242,239,230,0.06)" },
          ticks: {
            color: PALETTE.creamDim,
            font: { family: "IBM Plex Mono", size: 11 },
            callback: (v) => fmtShort(v),
          },
          suggestedMax: ceiling * 1.1,
        },
      },
    },
  });
}

function renderCategoryChart(d) {
  const ctx = document.getElementById("chart-category");
  const labels = d.category_breakdown.map(r => r.category);
  const values = d.category_breakdown.map(r => r.revenue_rwf);
  const colors = labels.map((_, i) => CATEGORY_COLORS[i % CATEGORY_COLORS.length]);

  new Chart(ctx, {
    type: "doughnut",
    data: { labels, datasets: [{ data: values, backgroundColor: colors, borderColor: "#0E3B36", borderWidth: 2 }] },
    options: {
      responsive: true,
      maintainAspectRatio: false,
      cutout: "62%",
      plugins: {
        legend: { display: false },
        tooltip: {
          callbacks: { label: (ctx) => `${ctx.label}: ${fmtRWF(ctx.parsed)}` },
        },
      },
    },
  });

  const total = values.reduce((a, b) => a + b, 0);
  const legend = document.getElementById("category-legend");
  legend.innerHTML = labels.map((label, i) => {
    const pct = Math.round((values[i] / total) * 100);
    return `<li><span class="swatch" style="background:${colors[i]}"></span>${label} · ${pct}%</li>`;
  }).join("");
}

function renderRegionChart(d) {
  const ctx = document.getElementById("chart-region");
  const sorted = [...d.region_breakdown].sort((a, b) => a.revenue_rwf - b.revenue_rwf);
  new Chart(ctx, {
    type: "bar",
    data: {
      labels: sorted.map(r => r.region),
      datasets: [{
        data: sorted.map(r => r.revenue_rwf),
        backgroundColor: PALETTE.sage,
        borderRadius: 4,
        maxBarThickness: 18,
      }],
    },
    options: {
      indexAxis: "y",
      responsive: true,
      maintainAspectRatio: false,
      plugins: {
        legend: { display: false },
        tooltip: { callbacks: { label: (ctx) => fmtRWF(ctx.parsed.x) } },
      },
      scales: {
        x: {
          grid: { color: "rgba(242,239,230,0.06)" },
          ticks: { color: PALETTE.creamDim, font: { family: "IBM Plex Mono", size: 10 }, callback: (v) => fmtShort(v) },
        },
        y: {
          grid: { display: false },
          ticks: { color: PALETTE.creamDim, font: { family: "Inter", size: 11 } },
        },
      },
    },
  });
}

function renderClientChart(d) {
  const ctx = document.getElementById("chart-client");
  const sorted = [...d.client_type_breakdown].sort((a, b) => b.revenue_rwf - a.revenue_rwf);
  new Chart(ctx, {
    type: "bar",
    data: {
      labels: sorted.map(r => r.client_type),
      datasets: [{
        data: sorted.map(r => r.revenue_rwf),
        backgroundColor: PALETTE.gold,
        borderRadius: 4,
        maxBarThickness: 32,
      }],
    },
    options: {
      responsive: true,
      maintainAspectRatio: false,
      plugins: {
        legend: { display: false },
        tooltip: { callbacks: { label: (ctx) => fmtRWF(ctx.parsed.y) } },
      },
      scales: {
        x: {
          grid: { display: false },
          ticks: { color: PALETTE.creamDim, font: { family: "Inter", size: 10 }, maxRotation: 0, autoSkip: false },
        },
        y: {
          grid: { color: "rgba(242,239,230,0.06)" },
          ticks: { color: PALETTE.creamDim, font: { family: "IBM Plex Mono", size: 10 }, callback: (v) => fmtShort(v) },
        },
      },
    },
  });
}

function renderTopProducts(d) {
  const rows = [...d.top_products].sort((a, b) => b.revenue - a.revenue);
  const max = rows[0]?.revenue || 1;
  const tbody = document.querySelector("#top-products-table tbody");
  tbody.innerHTML = rows.map(r => `
    <tr>
      <td>${r.product}</td>
      <td class="num">${r.units.toLocaleString("en-US")}</td>
      <td class="num">${fmtRWF(r.revenue)}</td>
      <td style="min-width:120px">
        <div class="bar-cell">
          <div class="bar-track"><div class="bar-fill" style="width:${Math.max(4, (r.revenue / max) * 100)}%"></div></div>
        </div>
      </td>
    </tr>
  `).join("");
}

function renderRecent(d) {
  const tbody = document.querySelector("#recent-table tbody");
  tbody.innerHTML = d.recent_transactions.map(t => `
    <tr>
      <td>${fmtDate(t.date)}</td>
      <td>${t.client_name}</td>
      <td>${t.product}</td>
      <td><span class="tag">${t.region.split(" - ").pop()}</span></td>
      <td class="num">${fmtRWF(t.line_total_rwf)}</td>
    </tr>
  `).join("");
}

(async function init() {
  try {
    const d = await loadData();
    renderKPIs(d);
    renderMonthlyChart(d);
    renderCategoryChart(d);
    renderRegionChart(d);
    renderClientChart(d);
    renderTopProducts(d);
    renderRecent(d);
    document.getElementById("last-updated").textContent =
      `· loaded ${new Date().toLocaleTimeString("en-GB")}`;
  } catch (err) {
    console.error("Failed to load dashboard data:", err);
    document.getElementById("connection-label").textContent =
      "Could not load data — check data.json or the API.";
  }
})();
